### PR TITLE
Fix rebuild for submit job

### PR DESCRIPTION
**Pull request by darionyaphet**

### CODE REVIEW

File: src/executor/admin/SubmitJobExecutor.cpp [hunk 1]
```diff
@@ -18,16 +18,25 @@ folly::Future<Status> SubmitJobExecutor::execute() {
 
     auto *sjNode = asNode<SubmitJob>(node());
     auto jobOp = sjNode->jobOp();
-    meta::cpp2::AdminCmd cmd = meta::cpp2::AdminCmd::COMPACT;
+    meta::cpp2::AdminCmd cmd;
     if (jobOp == meta::cpp2::AdminJobOp::ADD) {
         std::vector<std::string> params;
         folly::split(" ", sjNode->params().front(), params, true);
```
Comment (jievince) [inline src/executor/admin/SubmitJobExecutor.cpp:24]: Only the first element of sjNode->params() is taken here, and the other elements are lost.

Comment (jievince) [inline src/executor/admin/SubmitJobExecutor.cpp:24]: Only the first element of sjNode->params() is taken here, and the other elements are lost.

File: src/executor/admin/SubmitJobExecutor.cpp [hunk 2]
```diff
@@ -18,16 +18,25 @@ folly::Future<Status> SubmitJobExecutor::execute() {
...
         if (params.front() == "compact") {
             cmd = meta::cpp2::AdminCmd::COMPACT;
         } else if (params.front() == "flush") {
             cmd = meta::cpp2::AdminCmd::FLUSH;
+        } else if (params.front() == "rebuild") {
+            if (params[1] == "tag") {
```
Comment (jievince) [inline src/executor/admin/SubmitJobExecutor.cpp:30]: "tag" and "edge" are lost

**Review by jievince — CHANGES_REQUESTED** [review]

**Review by yixinglu — APPROVED** [review]

File: src/parser/parser.yy [hunk 3]
```diff
@@ -1866,8 +1876,10 @@ admin_job_sentence
     ;
 
 admin_job_operation
-    : KW_COMPACT { $$ = new std::string("compact"); }
-    | KW_FLUSH   { $$ = new std::string("flush"); }
+    : KW_COMPACT         { $$ = new std::string("compact"); }
```
Comment (laura-ding) [inline src/parser/parser.yy:1879]: you can directly to use `AdminCmd`

File: src/executor/admin/SubmitJobExecutor.cpp [hunk 4]
```diff
@@ -18,20 +18,35 @@ folly::Future<Status> SubmitJobExecutor::execute() {
...
+    std::vector<std::string> jobArguments;
+    auto spaceName = qctx()->rctx()->session()->space().name;
+    jobArguments.emplace_back(std::move(spaceName));
+
+    meta::cpp2::AdminCmd cmd;
     if (jobOp == meta::cpp2::AdminJobOp::ADD) {
-        std::vector<std::string> params;
-        folly::split(" ", sjNode->params().front(), params, true);
-        if (params.front() == "compact") {
+        auto& params = sjNode->params();
```
Comment (laura-ding) [inline src/executor/admin/SubmitJobExecutor.cpp:27]: if the parser directly to use `AdminCmd`, the converted code is not needed.

Comment (darionyaphet) [reply src/executor/admin/SubmitJobExecutor.cpp:27]: maybe I  will fix it later BTW would support REBUILD sentence in V2.0 ?

File: src/parser/parser.yy (hunk 3 above)
Comment (darionyaphet) [reply src/parser/parser.yy:1879]: seems Submit command treat the command type as String

File: src/executor/admin/SubmitJobExecutor.cpp (hunk 4 above)
Comment (laura-ding) [reply src/executor/admin/SubmitJobExecutor.cpp:27]: I think it's not needed, and we need to modify the doc.

File: src/executor/admin/SubmitJobExecutor.cpp [hunk 5]
```diff
@@ -18,20 +18,35 @@ folly::Future<Status> SubmitJobExecutor::execute() {
...
-        } else if (params.front() == "flush") {
+        } else if (params[0] == "flush") {
             cmd = meta::cpp2::AdminCmd::FLUSH;
+        } else if (params[0] == "rebuild") {
+            if (params[1] == "tag") {
+                cmd = meta::cpp2::AdminCmd::REBUILD_TAG_INDEX;
+                jobArguments.emplace_back(params[2]);
+            } else if (params[1] == "edge") {
+                cmd = meta::cpp2::AdminCmd::REBUILD_EDGE_INDEX;
+                jobArguments.emplace_back(params[2]);
```
Comment (jievince) [inline src/executor/admin/SubmitJobExecutor.cpp:38]: params[3] is index name

File: src/executor/admin/SubmitJobExecutor.cpp [hunk 6]
```diff
@@ -18,20 +18,35 @@ folly::Future<Status> SubmitJobExecutor::execute() {
 
     auto *sjNode = asNode<SubmitJob>(node());
     auto jobOp = sjNode->jobOp();
-    meta::cpp2::AdminCmd cmd = meta::cpp2::AdminCmd::COMPACT;
+    std::vector<std::string> jobArguments;
+    auto spaceName = qctx()->rctx()->session()->space().name;
+    jobArguments.emplace_back(std::move(spaceName));
+
+    meta::cpp2::AdminCmd cmd;
```
Comment (laura-ding) [inline src/executor/admin/SubmitJobExecutor.cpp:25]: compiler error does not init cmd

File: src/executor/admin/SubmitJobExecutor.cpp [hunk 7]
```diff
@@ -18,20 +18,35 @@ folly::Future<Status> SubmitJobExecutor::execute() {
...
+        auto& params = sjNode->params();
+        if (params[0] == "compact") {
             cmd = meta::cpp2::AdminCmd::COMPACT;
-        } else if (params.front() == "flush") {
+        } else if (params[0] == "flush") {
             cmd = meta::cpp2::AdminCmd::FLUSH;
+        } else if (params[0] == "rebuild") {
+            if (params[1] == "tag") {
+                cmd = meta::cpp2::AdminCmd::REBUILD_TAG_INDEX;
+                jobArguments.emplace_back(params[3]);
```
Comment (laura-ding) [inline src/executor/admin/SubmitJobExecutor.cpp:35]: I think the handle needs to move to validator because needs to check the index if exists.

**Review by jievince — APPROVED** [review]

File: src/executor/admin/SubmitJobExecutor.cpp [hunk 8]
```diff
@@ -18,20 +18,35 @@ folly::Future<Status> SubmitJobExecutor::execute() {
 
     auto *sjNode = asNode<SubmitJob>(node());
     auto jobOp = sjNode->jobOp();
-    meta::cpp2::AdminCmd cmd = meta::cpp2::AdminCmd::COMPACT;
+    std::vector<std::string> jobArguments;
+    auto spaceName = qctx()->rctx()->session()->space().name;
+    jobArguments.emplace_back(std::move(spaceName));
```
Comment (jievince) [inline src/executor/admin/SubmitJobExecutor.cpp:23]: jobArguments only need spaceName for AdminCmd::Add.

File: src/executor/admin/SubmitJobExecutor.cpp [hunk 9]
```diff
@@ -18,20 +18,35 @@ folly::Future<Status> SubmitJobExecutor::execute() {
...
+
+    meta::cpp2::AdminCmd cmd;
     if (jobOp == meta::cpp2::AdminJobOp::ADD) {
```
Comment (jievince) [inline src/executor/admin/SubmitJobExecutor.cpp:26]: The parms in sjNode->params() is missing for those who are not AdminJobOp::ADD.

**Review by laura-ding — DISMISSED** [review]

**Review by jievince — APPROVED** [review]

**Comment by laura-ding** [comment]
@darionyaphet detected memory leaks, please fix it

File: src/parser/parser.yy [hunk 10]
```diff
@@ -1520,13 +1520,23 @@ describe_edge_index_sentence
 
 rebuild_tag_index_sentence
     : KW_REBUILD KW_TAG KW_INDEX name_label {
-        $$ = new RebuildTagIndexSentence($4, false);
+        auto sentence = new AdminJobSentence(meta::cpp2::AdminJobOp::ADD);
+        sentence->addPara("rebuild");
+        sentence->addPara("tag");
+        sentence->addPara("index");
+        sentence->addPara(*$4);
```
Comment (laura-ding) [inline src/parser/parser.yy:1527]: need to add `delete $4;`

File: src/parser/parser.yy [hunk 11]
```diff
@@ -1520,13 +1520,23 @@ describe_edge_index_sentence
...
     ;
 
 rebuild_edge_index_sentence
     : KW_REBUILD KW_EDGE KW_INDEX name_label {
-        $$ = new RebuildEdgeIndexSentence($4, false);
+        auto sentence = new AdminJobSentence(meta::cpp2::AdminJobOp::ADD);
+        sentence->addPara("rebuild");
+        sentence->addPara("edge");
+        sentence->addPara("index");
+        sentence->addPara(*$4);
```
Comment (laura-ding) [inline src/parser/parser.yy:1538]: need to add `delete $4;`

**Review by laura-ding — APPROVED** [review]

**Review by jievince — APPROVED** [review]

**Pull request merged**